### PR TITLE
Run utility analysis for many parameters at once

### PR DESCRIPTION
**Pull request by dvadym**
This PR implements possibility to run utility analysis for multiple parameter at once, which is significant performance improvements in comparison to run a separate pipeline for each parameter set.

It contains:
1. `class MultiParameterConfiguration:` (in **dp_engine.py**): for specifying multi parameters run and generating `AggregateParams` for each of the parameter set.
2. `UtilityAnalysisEngine.aggregate` supports multi paramers run:
     a. it takes `MultiParameterConfiguration` (API)
     b. it creates combiners for each parameter set (implementation)
3. `perform_utility_analysis` (**utility_analysis.py**)  supports multi paramers run:
     a. it takes `MultiParameterConfiguration` (API)
     b. it creates combiners for each parameter set (implementation)

### CODE REVIEW

File: pipeline_dp/aggregate_params.py [hunk 1]
```diff
@@ -63,7 +63,7 @@ class PartitionSelectionStrategy(Enum):
 class AggregateParams:
     """Specifies parameters for function DPEngine.aggregate()
 
-    Args:
+    Attributes:
```
Comment (dvadym) [inline pipeline_dp/aggregate_params.py:66]: Fixes according to styleguide

File: utility_analysis_new/combiners.py [hunk 2]
```diff
@@ -111,8 +111,9 @@ def compute_probability_to_keep(self,
         otherwise it is an approximation computed from self.moments.
         """
         pmf = self.compute_pmf()
-        ps_strategy = partition_selection.create_truncated_geometric_partition_strategy(
-            eps, delta, max_partitions_contributed)
+        ps_strategy = partition_selection.create_partition_selection_strategy(
```
Comment (dvadym) [inline utility_analysis_new/combiners.py:114]: This was an incorrect merge in the previous PR.

File: examples/restaurant_visits/run_without_frameworks_utility_analysis.py [hunk 3]
```diff
@@ -31,27 +32,21 @@
 flags.DEFINE_string('output_file', None, 'Output file')
 flags.DEFINE_boolean('public_partitions', False,
                      'Whether public partitions are used')
+flags.DEFINE_boolean(
+    'per_partitions_metrics', False,
+    'Whether per partition or aggregate utility analysis is computed')
```
Comment (Kulankhina) [inline examples/restaurant_visits/run_without_frameworks_utility_analysis.py:37]: Does this refer to different types of utility analysis: per partition and aggregated? Just for my understanding, what would be an example of each?

File: utility_analysis_new/dp_engine.py [hunk 4]
```diff
@@ -21,6 +25,70 @@
...
+class MultiParameterConfiguration:
+    """Specifies parameters for multi-parameter Utility Analysis.
+
+    All MultiParameterConfiguration attributes corresponds to attributes in
+    pipeline_dp.AggregateParams.
+
+    UtilityAnalysisEngine can perform utility analysis for multiple sets of
+    parameters simultaneously. API for this is the following:
+    1. Specify blue-print AggregateParams instance.
+    2. Set MultiParameterConfiguration attributes which should have different
```
Comment (Kulankhina) [inline utility_analysis_new/dp_engine.py:38]: Maybe just "Set MultiParameterConfiguration attributes (see the example below). Note that each attribute is a sequence of parameter values for which the utility analysis will be run. All attributes that have non-None values must have the same sequence size."?
Otherwise, I am not sure what "which should have different values" means in this context

File: utility_analysis_new/dp_engine.py [hunk 5]
```diff
@@ -21,6 +25,70 @@
...
+    values.
+    3. Pass created objects to UtilityAnalysisEngine.aggregate().
```
Comment (Kulankhina) [inline utility_analysis_new/dp_engine.py:40]: nit: pass the created

File: utility_analysis_new/dp_engine.py [hunk 6]
```diff
@@ -21,6 +25,70 @@
...
+          AggregateParams(max_partitions_contributed=1, max_contributions_per_partition=10)
+          AggregateParams(max_partitions_contributed=2, max_contributions_per_partition=11)
+    """
+    max_partitions_contributed: Sequence[int] = None
+    max_contributions_per_partition: Sequence[int] = None
+    min_sum_per_partition: Sequence[float] = None
+    max_sum_per_partition: Sequence[float] = None
+
+    def __post_init__(self):
+        field_values = dataclasses.asdict(self)
```
Comment (Kulankhina) [inline utility_analysis_new/dp_engine.py:56]: nit: using the same notation ("attribute"-s or "field"-s) would probably make it slightly easier to read.

File: utility_analysis_new/dp_engine.py [hunk 7]
```diff
@@ -29,16 +97,21 @@ def __init__(self, budget_accountant: budget_accounting.BudgetAccountant,
...
+        public_partitions=None,
+        multi_param_configuration: Optional[MultiParameterConfiguration] = None
+    ):
         _check_utility_analysis_params(params, public_partitions)
         self._is_public_partitions = public_partitions is not None
+        self._multi_run_configuration = multi_param_configuration
         result = super().aggregate(col, params, data_extractors,
                                    public_partitions)
         self._is_public_partitions = None
+        self._multi_run_configuration = None
```
Comment (Kulankhina) [inline utility_analysis_new/dp_engine.py:114]: Why is this needed?

File: utility_analysis_new/dp_engine.py [hunk 8]
```diff
@@ -29,16 +97,21 @@ def __init__(self, budget_accountant: budget_accounting.BudgetAccountant,
...
 
-    def aggregate(self,
-                  col,
-                  params: pipeline_dp.AggregateParams,
-                  data_extractors: pipeline_dp.DataExtractors,
-                  public_partitions=None):
+    def aggregate(
+        self,
+        col,
+        params: pipeline_dp.AggregateParams,
```
Comment (Kulankhina) [inline utility_analysis_new/dp_engine.py:103]: Is this API public-facing?
If yes, then I believe it would be nice to explain the relation between params and multi_param_configuration. E.g., if multi_param_configuration has some parameters set, what should be the values of the corresponding parameters in params?

File: utility_analysis_new/tests/dp_engine_test.py [hunk 9]
```diff
@@ -19,6 +19,66 @@
...
 from pipeline_dp import budget_accounting
 from utility_analysis_new import dp_engine
+from utility_analysis_new.combiners import CountUtilityAnalysisMetrics
+import utility_analysis_new
+
+
+class MultiParameterConfiguration(parameterized.TestCase):
+
+    @parameterized.named_parameters(
+        dict(testcase_name="negative max_partitions_contributed",
```
Comment (Kulankhina) [inline utility_analysis_new/tests/dp_engine_test.py:29]: "All MultiParameterConfiguration fields unset"

File: utility_analysis_new/tests/dp_engine_test.py [hunk 10]
```diff
@@ -19,6 +19,66 @@
...
+             error_msg="MultiParameterConfiguration can not be empty.",
```
Comment (Kulankhina) [inline utility_analysis_new/tests/dp_engine_test.py:30]: nit/opt: I would say "MultiParameterConfiguration must have at least 1 non-empty attribute" to avoid confusion on what "empty MultiParameterConfiguration" might mean (now and if we amend the data structure in the future).

File: utility_analysis_new/tests/dp_engine_test.py [hunk 11]
```diff
@@ -167,6 +227,83 @@ def test_aggregate_error_metrics(self):
...
+        budget_accountant = pipeline_dp.NaiveBudgetAccountant(total_epsilon=1,
+                                                              total_delta=1e-10)
+
+        engine = dp_engine.UtilityAnalysisEngine(
+            budget_accountant=budget_accountant,
+            backend=pipeline_dp.LocalBackend())
+
+        # Input collection has 1 privacy id, which contributes to 2 partitions
+        # 1 and 2 times correspondingly.
+        input = [(0, 0), (0, 1), (0, 1)]
```
Comment (Kulankhina) [inline utility_analysis_new/tests/dp_engine_test.py:251]: very nit: why don't you use "pk0" and "pk1" as partition keys here but transform them with extractors? This seems to be a bit more straightforward.

File: examples/restaurant_visits/run_without_frameworks_utility_analysis.py (hunk 3 above)
Comment (dvadym) [reply examples/restaurant_visits/run_without_frameworks_utility_analysis.py:37]: Yes.
Cross-partition aggregated example:
```
(PartitionSelectionMetrics(num_partitions=7, dropped_partitions_expected=0.0, dropped_partitions_variance=0.0), 
 AggregateErrorMetrics(abs_error_expected=-619.9999999999998, abs_error_variance=131.3383283342634, abs_error_quantiles=[-605.3363229857413, -619.9999999999998, -634.6636770142583, -646.618370079588], rel_error_expected=-0.7623124338624336, rel_error_variance=0.000213443176382021, rel_error_quantiles=[-0.7437896396105713, -0.7623124338624336, -0.7808352281142958, -0.7959361008202313]))
```

Per partition:
```
(1, (1.0, CountUtilityAnalysisMetrics(count=600, per_partition_error=-300, expected_cross_partition_error=-162.80000000000007, std_cross_partition_error=7.645187302407234, std_noise=8.3515625, noise_kind=<NoiseKind.GAUSSIAN: 'gaussian'>)))
(2, (1.0, CountUtilityAnalysisMetrics(count=600, per_partition_error=-300, expected_cross_partition_error=-161.3666666666666, std_cross_partition_error=7.506737714287814, std_noise=8.3515625, noise_kind=<NoiseKind.GAUSSIAN: 'gaussian'>)))
... # other partitions
```
i.e. per-partition contains metrics per each partition

File: utility_analysis_new/dp_engine.py (hunk 4 above)
Comment (dvadym) [reply utility_analysis_new/dp_engine.py]: Thanks! I like that explanation.

File: utility_analysis_new/tests/dp_engine_test.py (hunk 9 above)
Comment (dvadym) [reply utility_analysis_new/tests/dp_engine_test.py]: Thans, copy-paste :)

File: utility_analysis_new/tests/dp_engine_test.py (hunk 10 above)
Comment (dvadym) [reply utility_analysis_new/tests/dp_engine_test.py]: Thanks, good point!

File: utility_analysis_new/tests/dp_engine_test.py (hunk 11 above)
Comment (dvadym) [reply utility_analysis_new/tests/dp_engine_test.py]: Good point! Done
I've updated in utility_analysis_test.py as well

File: utility_analysis_new/dp_engine.py (hunk 8 above)
Comment (dvadym) [reply utility_analysis_new/dp_engine.py:106]: I've added docstring for this function. PTAL

**Review by dvadym — COMMENTED** [review]
Thanks for the comments! I've addressed them. PTAL

File: examples/restaurant_visits/run_without_frameworks_utility_analysis.py (hunk 3 above)
Comment (Kulankhina) [reply examples/restaurant_visits/run_without_frameworks_utility_analysis.py:37]: I see, thanks for clarifying!

File: utility_analysis_new/dp_engine.py (hunk 7 above)
Comment (Kulankhina) [reply utility_analysis_new/dp_engine.py:136]: I see, thanks for clarifying!

File: utility_analysis_new/dp_engine.py (hunk 8 above)
Comment (Kulankhina) [reply utility_analysis_new/dp_engine.py:106]: Aaaaah, this is so helpful, thank you!

**Review by Kulankhina — APPROVED** [review]
